admin_manual/installation/nginx.rst

```diff
@@ -114,75 +76,109 @@ webroot of your nginx installation. In this example it is
+      add_header X-Robots-Tag                         "none"          always;
+      add_header X-XSS-Protection                     "1; mode=block" always;
+      
+      # Remove X-Powered-By, which is an information leak
+      fastcgi_hide_header X-Powered-By;
+      
+      # Path to the root of your installation
+      root /var/www/nextcloud;
+      
+      # Specify how to handle directories -- specifying `/index.php$request_uri`
+      # here as the fallback means that Nginx always exhibits the desired behaviour
+      # when a client requests a path that corresponds to a directory that exists
+      # on the server. In particular, if that directory contains an index.php file,
+      # that file is correctly served; if it doesn't, then the request is passed to
+      # the front-end controller. This consistent behaviour means that we don't need
+      # to specify custom rules for certain paths (e.g. images and other assets,
+      # `/updater`, `/ocm-provider`, `/ocs-provider`), and thus
+      # `try_files $uri $uri/ /index.php$request_uri`
+      # always provides the desired behaviour.
+      index index.php index.html index.htm /index.php$request_uri;
+      
+      # Default Cache-Control policy
+      expires 1m;
+      
+      # Rule borrowed from `.htaccess` to handle Microsoft DAV clients
+      if ( $http_user_agent ~ DavClnt ) { rewrite ^/$ /remote.php/webdav/ redirect; }
+      
+      location = /robots.txt {
+          allow all;
+          log_not_found off;
+          access_log off;
       }
-      location ~ ^\/(?:\.|autotest|occ|issue|indie|db_|console) {
-          deny all;
+      
+      # Make a regex exception for `/.well-known` so that clients can still
+      # access it despite the existence of the regex rule `location ~ /\.`
+      # which would otherwise handle requests for `/.well-known`.
+      location ^~ /.well-known {
+          # The following 6 rules are borrowed from `.htaccess`
+      
+          rewrite ^/\.well-known/host-meta\.json  /public.php?service=host-meta-json  last;
+          rewrite ^/\.well-known/host-meta        /public.php?service=host-meta       last;
+          rewrite ^/\.well-known/webfinger        /public.php?service=webfinger       last;
+          rewrite ^/\.well-known/nodeinfo         /public.php?service=nodeinfo        last;
+          
+          location /.well-known/carddav   { return 301 /remote.php/dav/; }
+          location /.well-known/caldav    { return 301 /remote.php/dav/; }
       }
-
-      location ~ ^\/(?:index|remote|public|cron|core\/ajax\/update|status|ocs\/v[12]|updater\/.+|oc[ms]-provider\/.+|.+\/richdocumentscode\/proxy)\.php(?:$|\/) {
-          fastcgi_split_path_info ^(.+?\.php)(\/.*|)$;
+      
+      # Rules borrowed from `.htaccess` to hide certain paths from clients
+      location ~ ^/(build|tests|config|lib|3rdparty|templates|data)($|/)  { return 404; }
+      location ~ ^/(autotest|occ|issue|indie|db_|console)                 { return 404; }
```

Comment:
`/.htaccess` uses the regex `^(?:build|tests|config|lib|3rdparty|templates)/.*`, which already includes `config`, so `/config/.htaccess` is already implemented by that regex. I added `data` to the end to implement `/data/.htaccess`.

Re non-capturing group `(?:`, it is my understanding that Nginx doesn't bother capturing a group to e.g. `$1` if `$1` does not appear anywhere within the `location` block, as is the case here — but I could be wrong about this. Even if Nginx does still internally capture the group, surely any performance gain is negligible?

I'm happy to let you decide whatever you prefer with regards to both of these points.

Reply:
NOTE: change `location ~ ^/(autotest|...` to `location ~ ^/(\.|autotest|...`

Reply:
I am fairly certain NGINX does capture text in parenthesis in a regex. If you search for the word "capture" on this page you can see an example. But feel free to test it.
http://nginx.org/en/docs/http/ngx_http_core_module.html

I think we should make it non-capturing because...
* There is no need to capture the blocked directory/file
* It is (negligibly) better performance
* It matches the `.htaccess`

I did not see "config" there, wasn't reading closely enough. I agree then, your order is best.

Reply:
I would also be interested in knowing what NGINX does different between `deny` and `return 404`. Not critical we figure that out now, just noting it down for future reference.

Reply:
`deny` returns a 403 🙂

Reply:
> I am fairly certain NGINX does capture text in parenthesis in a regex.

It does, but I believe that it is intelligent enough to render capturing groups as non-capturing groups if no captures are used in the block, i.e. no instances of `$1, `$2`, etc. appear. That is, when the config is loaded, Nginx will essentially rewrite the group internally to use `(?:` rather than `(`. I'm not completely sure of this (can't find any documentation on it now, though I'm pretty sure I read it somewhere in the past), and no harm in using non-capturing, so I'll change it

Reply:
That optimization makes sense to me based on how NGINX renders configs, but as you said since it's not documented it makes some sense to use the non-capturing version, that matches apache. I think this still isn't changed (back) in your current (as of now) version, do you mind changing it so I can 'resolve' this conversation? Thanks!

So sorry about the delay reviewing. I didn't have my test server setup at my new home. Finally just spun up a VM, and hope to get through the remaining notes ASAP.

Reply:
closing since I made the changes already